climpred/relative_entropy.py

```diff
@@ -0,0 +1,291 @@
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import xarray as xr
+from eofs.xarray import Eof
+
+from climpred.prediction import _pseudo_ens
+
+
+def _relative_entropy_formula(sigma_b, sigma_x, mu_x, mu_b, neofs):
+    """
+    Computes the relative entropy formula given in Branstator and Teng, (2010).
+
+    References:
+        - Branstator, Grant, and Haiyan Teng. “Two Limits of Initial-Value
+            Decadal Predictability in a CGCM.” Journal of Climate 23, no. 23
+            (August 27, 2010): 6292–6311. https://doi.org/10/bwq92h.
+        - Kleeman, Richard. “Measuring Dynamical Prediction Utility Using
+            Relative Entropy.” Journal of the Atmospheric Sciences 59, no. 13
+            (July 1, 2002): 2057–72. https://doi.org/10/fqwxpk.
+
+    Args:
+        sigma_b (xr.DataArray): covariance matrix of baseline distribution
+        sigma_x (xr.DataArray): covariance matrix of forecast distribution
+        mu_b (xr.DataArray): mean state vector of the baseline distribution
+        mu_x (xr.DataArray): mean state vector of the forecast distribution
+        neofs (int): number of EOFs used
+
+    Returns:
+        R (float): relative entropy
+        dispersion (float): dispersion component
+        signal (float): signal component
+    """
+    fac = 0.5
+    dispersion = fac * (np.log(np.linalg.det(sigma_b) / np.linalg.det(sigma_x))
+                        + np.trace(sigma_x / sigma_b) - neofs)
+
+    # https://stackoverflow.com/questions/7160162/left-matrix-division-and-numpy-solve
+    # (A.T\B)*A
+    x, resid, rank, s = np.linalg.lstsq(
+        sigma_b, mu_x - mu_b)  # sigma_b \ (mu_x - mu_b)
+    signal = fac * np.matmul((mu_x.values - mu_b.values), x)
+    R = dispersion + signal
+    return R, dispersion, signal
+
+
+def _gen_control_distribution(ds, control, it=10):
+    """
+    Generate a large control distribution from control.
+
+    Args:
+        ds (xr.DataArray): initialization data with dimensions initialization,
+                           member, time and spatial [lon (x), lat(y)].
+        control (xr.DataArray): control data with dimensions time and spatial
+                                [lon (x), lat(y)].
+        it (int): multiplying factor for ds.member.
+
+    Returns:
+        control_uninitialized (xr.DataArray): data with dimensions
+                                              initialization, member, time
+                                              and spatial [lon (x), lat(y)].
+
+    """
+    ds_list = []
+    for _ in range(it):
+        control_uninitialized = _pseudo_ens(ds, control)
+        control_uninitialized['initialization'] = ds.initialization.values
+        ds_list.append(control_uninitialized)
+    control_uninitialized = xr.concat(ds_list, 'member')
+    control_uninitialized['member'] = np.arange(control_uninitialized.member.size)
+    return control_uninitialized
+
+
+def _gen_control_distribution_lens(ds, control, max_member=40):
```

Comment:
What's the point of this function? It's virtually the same as `_gen_control_distribution()` but just constrains the number of members differently. This could just be a `lens=True/False` flag in `_gen_control_distribution()` if it's a needed function at all.

Reply:
Sorry. Yes they are the same. I omitted the _lens function. Shouldn’t have been submitted. before Skyping I will think of a proposed structure of dple/lens and perfect model.